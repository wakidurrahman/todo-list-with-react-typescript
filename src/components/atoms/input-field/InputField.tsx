/**
 * InputField Component
 *
 * A reusable form input component that supports both text input and textarea elements.
 * Follows Bootstrap styling conventions and implements accessibility best practices.
 *
 * Features:
 * - Support for `input` and `textarea` elements
 * - Built-in validation states and feedback
 * - Help text and error message display
 * - Accessibility support with ARIA attributes
 * - Bootstrap form styling
 * - Customizable size and appearance
 *
 * @example
 * <InputField
 *   id="email"
 *   label="Email Address"
 *   type="email"
 *   value={email}
 *   onChange={handleEmailChange}
 *   required
 *   placeholder="Enter your email"
 * />
 */

import { ChangeEvent } from 'react';

type InputSize = 'sm' | 'lg' | undefined;

type InputFieldProps = {
  tagName?: 'input' | 'textarea';
  /** Input id - required for accessibility */
  id: string;
  /** Input label */
  label?: string;
  /** Input type (text, email, password, etc.) */
  type?: 'text' | 'email' | 'password' | 'number' | 'search';
  /** Input value */
  value: string;
  /** Callback when input value changes */
  onChange: (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
  /** Input name attribute */
  name?: string;
  /** Input placeholder text */
  placeholder?: string;
  /** Additional CSS classes */
  className?: string;
  /** Input size - sm or lg */
  size?: InputSize;
  /** Whether the input is disabled */
  disabled?: boolean;
  /** Whether the input is readonly */
  readOnly?: boolean;
  /** Help text below the input */
  helpText?: string;
  /** Whether the input is required */
  required?: boolean;
  /** Error message */
  error?: string;
  /** Aria-describedby for accessibility */
  'aria-describedby'?: string;
  /** Whether the input has been touched */
  touched?: boolean;
  /** Maximum length of the input */
  maxLength?: number;
  /** Whether the input is valid */
  isValid?: boolean;
  /** Feedback message for valid input */
  validFeedback?: string;
  /** Aria-label for accessibility */
  'aria-label'?: string;
  /** Whether the input is tabbable */
  tabIndex?: number;
  /** Number of rows for textarea */
  rows?: number;
};

/**
 * InputField component that renders either an input or textarea element with associated label,
 * help text, and validation feedback.
 */
const InputField = ({
  tagName = 'input',
  type = 'text',
  value,
  onChange,
  placeholder,
  label,
  id,
  size,
  disabled = false,
  readOnly = false,
  helpText,
  required = false,
  name,
  error,
  className = '',
  touched = false,
  maxLength,
  isValid,
  validFeedback = 'Looks good!',
  'aria-describedby': ariaDescribedby,
  'aria-label': ariaLabel,
  tabIndex = 0,
  rows = 3,
}: InputFieldProps) => {
  // Determine whether to render an input or textarea element
  const Tag = tagName === 'input' ? 'input' : 'textarea';

  // Generate unique IDs for accessibility elements
  const helpTextId = `${id}-help-text`;
  const errorId = `${id}-error`;
  const feedbackId = `${id}-field`;

  // Combine aria-describedby values for accessibility
  const ariaDescribedbyValues = [
    helpText ? helpTextId : '',
    error ? errorId : '',
    ariaDescribedby || '',
    touched ? feedbackId : '',
  ]
    .filter(Boolean)
    .join(' ');

  // Build class names for form control with validation states
  const formControlClasses = [
    'form-control',
    size && `form-control-${size}`,
    touched && (error ? 'is-invalid' : isValid && 'is-valid'),
    className,
  ]
    .filter(Boolean)
    .join(' ');

  return (
    <div className="mb-3">
      {label && (
        <label htmlFor={id} className="form-label">
          {label}
          {required && <span className="text-danger ms-1">*</span>}
        </label>
      )}
      <Tag
        type={type}
        className={formControlClasses}
        id={id}
        name={name || id}
        value={value}
        onChange={onChange}
        placeholder={placeholder}
        disabled={disabled}
        readOnly={readOnly}
        required={required}
        maxLength={maxLength}
        aria-describedby={ariaDescribedbyValues || undefined}
        aria-label={ariaLabel}
        tabIndex={tabIndex}
        rows={rows}
      />

      {/* Help text */}
      {helpText && (
        <div id={helpTextId} className="form-text">
          {helpText}
        </div>
      )}

      {/* Error message */}
      {touched && error && (
        <div id={errorId} className="invalid-feedback">
          {error}
        </div>
      )}

      {/* Valid feedback */}
      {touched && !error && isValid && (
        <div id={feedbackId} className="valid-feedback">
          {validFeedback}
        </div>
      )}
    </div>
  );
};

export default InputField;
