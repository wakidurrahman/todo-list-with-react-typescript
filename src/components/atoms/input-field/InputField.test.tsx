import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import InputField from './InputField';

/**
 * Test suite for the InputField component
 *
 * Categories:
 * - Basic Rendering
 *   - Default input rendering
 *   - Textarea variant
 *   - Label rendering
 *   - Required field indicator
 *
 * - Input Behavior
 *   - Value changes
 *   - Disabled state
 *   - Readonly state
 *   - MaxLength validation
 *
 * - Validation and Feedback
 *   - Error messages
 *   - Valid feedback
 *   - Help text
 *   - Validation classes
 *
 * - Accessibility
 *   - ARIA attributes
 *   - Help text association
 *   - Tab index
 *   - Label associations
 *
 * - Styling
 *   - Size variants
 *   - Custom classes
 *
 * - Snapshots
 *   - Basic input
 *   - Textarea variant
 *   - Input with validation
 */
describe('InputField Component', () => {
  const defaultProps = {
    id: 'test-input',
    value: '',
    onChange: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  // Basic rendering tests
  describe('Basic Rendering', () => {
    it('renders input field with default props', () => {
      render(<InputField {...defaultProps} />);
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('renders textarea when tagName is textarea', () => {
      render(<InputField {...defaultProps} tagName="textarea" />);
      const textarea = screen.getByRole('textbox');
      expect(textarea.tagName.toLowerCase()).toBe('textarea');
    });

    it('renders label when provided', () => {
      const label = 'Test Label';
      render(<InputField {...defaultProps} label={label} />);
      expect(screen.getByLabelText(label)).toBeInTheDocument();
    });

    it('shows required asterisk when required prop is true', () => {
      render(<InputField {...defaultProps} label="Test Label" required />);
      expect(screen.getByText('*')).toHaveClass('text-danger');
    });
  });

  // Input behavior tests
  describe('Input Behavior', () => {
    it('handles value changes', () => {
      const handleChange = jest.fn();
      render(<InputField {...defaultProps} onChange={handleChange} />);

      const input = screen.getByRole('textbox');
      fireEvent.change(input, { target: { value: 'test value' } });

      expect(handleChange).toHaveBeenCalledTimes(1);
    });

    it('respects disabled state', () => {
      render(<InputField {...defaultProps} disabled />);
      expect(screen.getByRole('textbox')).toBeDisabled();
    });

    it('respects readonly state', () => {
      render(<InputField {...defaultProps} readOnly />);
      expect(screen.getByRole('textbox')).toHaveAttribute('readonly');
    });

    it('applies maxLength constraint', () => {
      const maxLength = 10;
      render(<InputField {...defaultProps} maxLength={maxLength} />);
      expect(screen.getByRole('textbox')).toHaveAttribute(
        'maxLength',
        maxLength.toString()
      );
    });
  });

  // Validation and feedback tests
  describe('Validation and Feedback', () => {
    it('shows error message when touched and has error', () => {
      const errorMessage = 'This field is required';
      render(
        <InputField {...defaultProps} touched={true} error={errorMessage} />
      );
      expect(screen.getByText(errorMessage)).toHaveClass('invalid-feedback');
    });

    it('shows valid feedback when touched and valid', () => {
      const validMessage = 'Looks good!';
      render(
        <InputField
          {...defaultProps}
          touched
          isValid
          validFeedback={validMessage}
        />
      );
      expect(screen.getByText(validMessage)).toHaveClass('valid-feedback');
    });

    it('shows help text when provided', () => {
      const helpText = 'Enter your username';
      render(<InputField {...defaultProps} helpText={helpText} />);
      expect(screen.getByText(helpText)).toHaveClass('form-text');
    });

    it('applies validation classes correctly', () => {
      render(<InputField {...defaultProps} touched error="Error" />);
      expect(screen.getByRole('textbox')).toHaveClass('is-invalid');
    });
  });

  // Accessibility tests
  describe('Accessibility', () => {
    it('associates help text with input via aria-describedby', () => {
      const helpText = 'Help text';
      render(<InputField {...defaultProps} helpText={helpText} />);

      const input = screen.getByRole('textbox');
      const helpTextElement = screen.getByText(helpText);

      expect(input).toHaveAttribute('aria-describedby', helpTextElement.id);
    });

    it('applies custom aria-label', () => {
      const ariaLabel = 'Custom label';
      render(<InputField {...defaultProps} aria-label={ariaLabel} />);
      expect(screen.getByRole('textbox')).toHaveAttribute(
        'aria-label',
        ariaLabel
      );
    });

    it('sets correct tabIndex', () => {
      const tabIndex = 2;
      render(<InputField {...defaultProps} tabIndex={tabIndex} />);
      expect(screen.getByRole('textbox')).toHaveAttribute(
        'tabindex',
        tabIndex.toString()
      );
    });
  });

  // Style tests
  describe('Styling', () => {
    it('applies size classes correctly', () => {
      render(<InputField {...defaultProps} size="lg" />);
      expect(screen.getByRole('textbox')).toHaveClass('form-control-lg');
    });

    it('applies custom className', () => {
      const customClass = 'custom-input';
      render(<InputField {...defaultProps} className={customClass} />);
      expect(screen.getByRole('textbox')).toHaveClass(customClass);
    });
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches basic input snapshot', () => {
      const tree = renderer
        .create(<InputField {...defaultProps} label="Basic Input" />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches textarea snapshot', () => {
      const tree = renderer
        .create(
          <InputField
            {...defaultProps}
            tagName="textarea"
            label="Textarea Input"
            rows={4}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches input with validation snapshot', () => {
      const tree = renderer
        .create(
          <InputField
            {...defaultProps}
            label="Validated Input"
            touched
            error="Error message"
            helpText="Help text"
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
