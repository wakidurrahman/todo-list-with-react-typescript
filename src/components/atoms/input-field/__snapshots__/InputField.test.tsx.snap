// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField Component Snapshots matches basic input snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-input"
  >
    Basic Input
  </label>
  <input
    className="form-control"
    disabled={false}
    id="test-input"
    name="test-input"
    onChange={[MockFunction]}
    readOnly={false}
    required={false}
    rows={3}
    tabIndex={0}
    type="text"
    value=""
  />
</div>
`;

exports[`InputField Component Snapshots matches input with validation snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-input"
  >
    Validated Input
  </label>
  <input
    aria-describedby="test-input-help-text test-input-error test-input-field"
    className="form-control is-invalid"
    disabled={false}
    id="test-input"
    name="test-input"
    onChange={[MockFunction]}
    readOnly={false}
    required={false}
    rows={3}
    tabIndex={0}
    type="text"
    value=""
  />
  <div
    className="form-text"
    id="test-input-help-text"
  >
    Help text
  </div>
  <div
    className="invalid-feedback"
    id="test-input-error"
  >
    Error message
  </div>
</div>
`;

exports[`InputField Component Snapshots matches textarea snapshot 1`] = `
<div
  className="mb-3"
>
  <label
    className="form-label"
    htmlFor="test-input"
  >
    Textarea Input
  </label>
  <textarea
    className="form-control"
    disabled={false}
    id="test-input"
    name="test-input"
    onChange={[MockFunction]}
    readOnly={false}
    required={false}
    rows={4}
    tabIndex={0}
    type="text"
    value=""
  />
</div>
`;
